src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,145 @@
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAddByIndex(index);
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            addByIndexZero(value);
+        } else {
+            addByIndexDefault(index, value);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return findNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = findNodeByIndex(index);
+        T oldValue = node.item;
+        node.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> node = findNodeByIndex(index);
+        T oldValue = node.item;
+        unLink(node);
+        size--;
+        return oldValue;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> node = head;
+        while (node != null) {
+            if (node.item == object || node.item != null && node.item.equals(object)) {
+                unLink(node);
+                size--;
+                return true;
+            }
+            node = node.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T item, Node<T> next) {
+            this.prev = prev;
+            this.item = item;
+            this.next = next;
+        }
+    }
+
+    private void checkIndexForAddByIndex(int index) {
+        if (index >= 0 && index <= size) {
+            return;
+        }
+        throw new IndexOutOfBoundsException("Cannot add element , because index "
+                + index + " is invalid");
+    }
+
+    private void addByIndexDefault(int index, T value) {
+        Node<T> nodeByIndex = findNodeByIndex(index);
+        Node<T> newNode = new Node<>(nodeByIndex.prev, value, nodeByIndex);
+        nodeByIndex.prev.next = newNode;
+        nodeByIndex.prev = newNode;
+        size++;
+    }
+
+    private void addByIndexZero(T value) {
+        Node<T> newNode = new Node<>(null, value, head);
+        head.prev = newNode;
+        head = newNode;
+        size++;
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " is invalid");
+        }
+        Node<T> node = head;
+        for (int i = 1; i <= index; i++) {
+            node = node.next;
+        }
```

Comment:
i think you need find index from head to half and from tail to half. its faster.